udffsck/udffsck.c

```diff
@@ -0,0 +1,3370 @@
+            return NULL;
+        case DSTRING_E_NONZERO_PADDING:
+            return "non-zero padding";
+        case DSTRING_E_WRONG_LENGTH:
+            return "wrong length";
+        case DSTRING_E_INVALID_CHARACTERS:
+            return "invalid characters present";
+        case DSTRING_E_NOT_EMPTY:
+            return "string is not empty";
+        case DSTRING_E_UNKNOWN_COMP_ID:
+            return "unknown Compression ID";
+        default: 
+            return "unknown dstring error";
+   } 
+}
+
+uint8_t dstring_error(char * string_name, uint8_t e_code) {
+    char * buffer[600];
+    if(e_code > 0) {
+        msg("Dstring %s has following errors:\n", string_name);
+        for(int i=0; i<8; ++i) {
+            if(e_code & 1<<i) {
+                msg("\t- %s\n", dstring_suberror(e_code & 1<<i));
+            }
+        }
+
+        return 4; 
+    }
+    return 0;
+}
+
+/**
+ * \brief Function for detection of errors in dstrings
+ *
+ * This function detects violation against UDF 2.1.1 in dstrings
+ * Theese violations are:
+ *  1. Non-zero padding
+ *  2. Not allowed characters for 16 bit dchars (0xFFFE and 0xFEFF)
+ *  3. Unknown compression ID
+ *  4. Emptyness of string if ID is 0 or length is 0
+ *  5. String legth mismatch
+ *
+ * \param[in] in dstring for check
+ * \param[in] field_size size of dstring field
+ * \return sum of DSTRING_E codes 
+ */
+uint8_t check_dstring(dstring *in, size_t field_size) {
+    uint8_t compID = in[0];
+    uint8_t length = in[field_size-1];
+    uint8_t stepping = 0xFF;
+    uint8_t empty_flag = 0;
+    uint8_t e_code = 0;
+
+    dbg("compID: %d, length: %d\n", compID, length);
+    switch(compID) {
+        case 8: 
+            stepping = 1;
+            break;
+        case 16: 
+            stepping = 2;
+            break;
+        case 0:
+            stepping = 1;
+            empty_flag = 1;
+            break;
+        default:
+            err("Unknown dstring compression ID.\n");
+            return DSTRING_E_UNKNOWN_COMP_ID;
+    }
+
+    if(empty_flag || length == 0) {
+        // Check for emptyness
+        for(int i = 0; i < field_size; i += stepping) {
+            if(in[i] != 0) {
+                err("Dstring is not empty.\n");
+                e_code |= DSTRING_E_NOT_EMPTY;
+            }
+        }
+    } else {
+        // Leave first and last bytes, they are special.
+        // Stepping +1 if 8bit or +2 if 16bit character length
+        
+        // Check for length and zero padding.
+        uint8_t char_count = 0;
+        uint8_t eol_flag = 0xFF;
+        for(int i = 1; i < field_size-1; i += stepping) {
+            // We need to check if character is 0.
+            // For 8bit: we check character twice to keep code simplicity.
+            // For 16bit: we check character i and i+1
+            // 
+            // If hole (NULL character) detected, eol_flag is set.
+            // If eol is set and characted is detected, it is violation of UDF 2.1.1
```

Comment:
Really? Currently I'm not able to find this restriction. U+0000 is a valid unicode code point, not disallowed by UDF and therefore I should be able to store e.g. U+0001 U+0000 U+0001 into dstring32 as: `0x08 0x01 0x00 0x01 ... 0x04`

Reply:
If I understand correctly to ECMA-167, 1/7.2.12, it should be this way.

> **7.2.12 Fixed-length character fields**
A dstring of length n is a field of n bytes where d-characters (1/7.2) are recorded. The number of bytes used to record the characters shall be recorded as a Uint8 (1/7.1.1) in byte n, where n is the length of the field. **The characters shall be recorded starting with the first byte of the field, and any remaining byte positions after the characters up until byte n − 1 inclusive shall be set to #00.**
Unless otherwise specified, a dstring shall not be all #00 bytes.

Correct me if I am wrong please.

Reply:
I understand it that sequence `0x08 0x01 0x00 0x01 ... 0x04` (where `...` are 27 zero bytes) is valid d-string. `n` is 32 bytes, `0x01 0x00 0x01` are valid d-characters (they represent valid code points U+0001 U+0000 U+0001). Padding is `0x00` bytes and I do not see where can be restriction that byte 0x00 is disallowed outside of padding.

Reply:
Ok, I understand your argument. But how can I detect correctness of declared length? Using this logic on your example, it can be anything from length 4 to 32.

Reply:
Ok, lets simplify it and disallow nulls. I think that U+00 would just cause lot of problems as char* in C is null term and even wchar_t* or char32_t* (both for handling unicode) is used as null term.

Reply:
Ok. So I'll keep it as is.